p2p/src/peers/mod.rs

```diff
@@ -92,9 +110,24 @@ impl Peers {
         // Note: if the peer address exists, the peer info will be overwritten
         let result = addrs
             .into_iter()
-            // Filter out unspecified addresses (aka 0.0.0.0)
-            .filter(|address| !address.ip().is_unspecified())
             .filter_map(|address| {
+                // Filter out unspecified addresses (aka 0.0.0.0), and the server address
+                if address.ip().is_unspecified()
+                    || self.is_server_address(&address).is_none()
+                    || self.is_server_address(&address).unwrap()
+                {
+                    return None;
+                }
+
+                let index = self.tried_bucket_index(&address);
+                let elem = self.tried_bucket.get(&index);
+
+                // If the index point to the same address that it is already
+                // in tried, we don't include in new bucket
+                if elem.is_some() && (elem.unwrap().address == address) {
```

Comment:
```suggestion
                if let Some(tried_address) = elem {
                    if address == tried_address {
                        return None
```

Reply:
to only have a single point of exit that you suggest after, it is more easy to evaluate this expression